### PR TITLE
Fix/reject gates targetting excluded positions

### LINKED ISSUE #412: Transpiler stage to reject gates targetting excluded positions
Given a QPU's connectivity, and it's excluded_positions, a transpiler stage should check that all the gates in a circuit are valid, i.e., they do not interact with any of the excluded positions. This transpiler stage should be added at the end of the Sequential Transpiler in  get_anyon_transpiler

### PR DESCRIPTION
**Pull request by npoirie1**
Adds a transpiler that throws an exception if a gate in the circuit is at an excluded position.

Fixes #412

### CODE REVIEW

File: src/core/transpile.jl [hunk 1]
```diff
@@ -2447,3 +2447,52 @@ function transpile(
 
     return circuit
 end
+
+struct RejectGatesOnExcludedPositionsTranspiler <: Transpiler
```
Comment (plerouxanyon) [inline src/core/transpile.jl:2451]: Please see other transpilers for how to add documentation and doctests to public functions.

File: src/core/transpile.jl [hunk 2]
```diff
@@ -2447,3 +2447,52 @@ function transpile(
...
+
+function are_gates_at_excluded_positions(
+    connectivity::Union{LineConnectivity,LatticeConnectivity},
+    circuit::QuantumCircuit,
+)::Tuple{Bool,String}
+
+    excluded_positions = get_excluded_positions(connectivity)
+    for instruction in get_circuit_instructions(circuit)
+        qubits = get_connected_qubits(instruction)
+        for (qubit, position) in Iterators.product(qubits, excluded_positions)
```
Comment (plerouxanyon) [inline src/core/transpile.jl:2486]: Minor, but for readability, let's change this to be (untested)

```julia
for qubit in qubits
  if qubit in excluded_positions
```

The above style is common in Python where the interpreter consumes CPU cycles in explicit for loops, in Julia there is usually no benefit to reducing nested levels of for loops.

File: test/transpile.jl [hunk 3]
```diff
@@ -1837,3 +1837,39 @@ end
...
+        qubit_count = 2,
+        instructions = [control_z(1, 2)],
+        name = "test-name",
+    )
+    transpiler = RejectGatesOnExcludedPositionsTranspiler(connectivity)
+
+    @test isequal(transpile(transpiler, circuit), circuit)
+end
+
+@testset "RejectGatesOnExcludedPositionsTranspiler for line connectivity" begin
```
[the remark this answers is not in the document]
Comment (npoirie1) [reply test/transpile.jl:1855]: The tests for lattice connectivity have been added.

File: src/core/transpile.jl (hunk 1 above)
Comment (npoirie1) [reply src/core/transpile.jl:2451]: Documentation and doctests have been added.

**Review by plerouxanyon — APPROVED** [review]

**Pull request merged**